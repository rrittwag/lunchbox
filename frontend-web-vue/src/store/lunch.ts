import { ref } from 'vue'
import api from '@/api/lunch'
import { ApiError } from '@/api/http'
import { LunchProvider } from '@/model/LunchProvider'
import { LunchOffer } from '@/model/LunchOffer'
import { today } from '@/util/date'
import { LunchLocation } from '@/model/LunchLocation'
import { defineStore } from 'pinia'

export const useLunchStore = defineStore('lunch', () => {
  // --------------------
  //  offers & providers
  // --------------------
  const offers = ref<LunchOffer[]>([])
  const providers = ref<LunchProvider[]>([])
  const isLoading = ref(false)
  const error = ref<ApiError | undefined>()

  async function loadFromApi() {
    isLoading.value = true
    try {
      const providersPromise = api.getProviders()
      const offersPromise = api.getOffers()
      providers.value = await providersPromise
      offers.value = await offersPromise
    } catch (e) {
      error.value = e as ApiError
    }
    isLoading.value = false
  }

  // -------------------
  //  locations
  // -------------------
  const locations: LunchLocation[] = [
    new LunchLocation('Neubrandenburg', 'NB'),
    new LunchLocation('Berlin Springpfuhl', 'B'),
  ]
  const selectedLocation = ref(loadSelectedLocationFromLocalStorage())

  function loadSelectedLocationFromLocalStorage(): LunchLocation {
    let locationName: string | null = localStorage.getItem('lunchboxWebapp.STORAGEKEY_LOCATION')
    if (!locationName || locationName === '') return locations[0]

    // die alte Angular-App speicherte den Wert mit ""
    locationName = locationName.replace(/"/g, '')

    const filteredLocation = locations.filter((l) => l.name === locationName)
    if (filteredLocation.length > 0) return filteredLocation[0]
    return locations[0]
  }

  function selectLocation(newLocation: LunchLocation) {
    selectedLocation.value = newLocation
    let locationName = ''
    if (newLocation) locationName = newLocation.name
    localStorage.setItem('lunchboxWebapp.STORAGEKEY_LOCATION', locationName)
  }

  // ------------------
  //  selected day
  // ------------------
  const selectedDay = ref(today())

  function selectDay(newSelectedDay: Date) {
    selectedDay.value = newSelectedDay
  }

  // ---------------------
  //  export
  // ---------------------
  return {
    offers,
    providers,
    isLoading,
    error,
    loadFromApi,
    locations,
    selectedLocation,
    selectLocation,
    selectedDay,
    selectDay,
  }
})
