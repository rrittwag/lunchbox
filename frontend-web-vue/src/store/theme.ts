import { computed, ref } from 'vue'

export interface Theme {
  cssClass: string
  label: string
}

export type ColorScheme = 'system' | 'light' | 'dark'

export const THEME_RED: Theme = { cssClass: 'theme-red', label: 'Red' }
export const THEME_GREEN: Theme = { cssClass: 'theme-green', label: 'Green' }
export const THEME_BLUE: Theme = { cssClass: 'theme-blue', label: 'Blue' }
export const THEME_BLUE_DARK: Theme = { cssClass: 'theme-blue-dark', label: 'Blue (Dark)' }

const currentTheme = ref<Theme | undefined>()
const colorScheme = ref<ColorScheme>('system')

export function useTheme() {
  // TODO: register listener for media query
  if (colorScheme.value === 'system')
    colorScheme.value = window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light'

  // TODO: Theme von ColorScheme lÃ¶sen?
  setCurrentTheme(colorScheme.value === 'dark' ? THEME_BLUE_DARK : THEME_RED)

  // TODO: Umstellen auf Tailwind Dark Mode 'class'? -> https://tailwindcss.com/docs/dark-mode
  function setCurrentTheme(newTheme?: Theme) {
    // background color works best on <body />, so theme must be set on <body />
    const bodyClasses = window.document.body.classList
    if (currentTheme.value) bodyClasses.remove(currentTheme.value.cssClass)
    if (newTheme) bodyClasses.add(newTheme.cssClass)

    // transition on theme change AFTER initial page load
    if (currentTheme.value && !bodyClasses.contains('theme-transition'))
      bodyClasses.add('theme-transition')

    currentTheme.value = newTheme
  }

  function setColorScheme(newScheme: ColorScheme) {
    colorScheme.value = newScheme
  }

  return {
    themes: computed(() => [THEME_RED, THEME_GREEN, THEME_BLUE, THEME_BLUE_DARK]),
    currentTheme: computed(() => currentTheme.value),
    setCurrentTheme,
    colorScheme: computed(() => colorScheme.value),
    setColorScheme,
  }
}

export function __reset__TEST_ONLY__() {
  currentTheme.value = undefined
}
