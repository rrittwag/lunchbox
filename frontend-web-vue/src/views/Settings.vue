<template>
  <div class="sm:p-4">
    <h1
      class="
        flex
        items-center
        justify-center
        sm:justify-start sm:h-16
        px-4
        text-2xl text-neutral-800
      "
    >
      Einstellungen
    </h1>
    <div>
      <div class="max-w-screen-xl mx-auto py-12 px-4 sm:px-6 lg:py-16 lg:px-8">
        <div class="mt-6 border-t-2 border-neutral-500 pt-6">
          <dl>
            <div class="md:grid md:grid-cols-12 md:gap-8">
              <dt class="text-base leading-6 font-medium text-neutral-900 md:col-span-5">Ort</dt>
              <dd class="mt-2 md:mt-0 md:col-span-7">
                <button
                  class="
                    rounded
                    px-4
                    h-20
                    border border-primary-500
                    text-base
                    leading-6
                    text-neutral-800
                  "
                >
                  Neubrandenburg
                </button>
                <button
                  class="
                    rounded
                    ml-2
                    px-4
                    h-20
                    border border-neutral-400
                    hover:border-neutral-800
                    text-base
                    leading-6
                    text-neutral-500
                  "
                >
                  Neubrandenburg
                </button>
                <button
                  class="
                    rounded
                    ml-2
                    px-4
                    h-20
                    border border-neutral-400
                    hover:border-neutral-800
                    text-base
                    leading-6
                    text-neutral-500
                  "
                >
                  Berlin Springpfuhl
                </button>
              </dd>
            </div>
            <div class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8">
              <dt
                class="text-base leading-6 font-medium text-neutral-900 md:col-span-5"
                for="settings-theme"
              >
                Farbvariante
              </dt>
              <dd class="mt-2 md:mt-0 md:col-span-7">
                <div class="inline-block relative w-32">
                  <select
                    id="settings-theme"
                    class="
                      block
                      appearance-none
                      w-full
                      bg-neutral-200
                      text-neutral-900
                      px-4
                      py-2
                      pr-8
                      rounded
                      leading-tight
                      border border-neutral-400
                      hover:border-neutral-800
                      focus:outline-none focus:ring-4
                    "
                    name="theme"
                    :value="currentTheme?.cssClass"
                    @change="onSelectTheme"
                  >
                    <option v-for="theme in themes" :key="theme.cssClass" :value="theme.cssClass">
                      {{ theme.label }}
                    </option>
                  </select>
                  <div
                    class="
                      pointer-events-none
                      absolute
                      inset-y-0
                      right-0
                      flex
                      items-center
                      px-2
                      text-neutral-900
                    "
                  >
                    <svg
                      class="fill-current h-4 w-4"
                      xmlns="http://www.w3.org/2000/svg"
                      viewBox="0 0 20 20"
                    >
                      <path
                        d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"
                      />
                    </svg>
                  </div>
                </div>
              </dd>
            </div>
            <div class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8">
              <dt class="text-base leading-6 font-medium text-neutral-900 md:col-span-5">
                Dark Mode
              </dt>
              <dd class="mt-2 md:mt-0 md:col-span-7">
                <p class="text-base leading-6 text-neutral-500">ja/nein/System</p>
              </dd>
            </div>
            <div class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8">
              <dt class="text-base leading-6 font-medium text-neutral-900 md:col-span-5">
                Mittagsanbieter anzeigen
              </dt>
              <dd class="mt-2 md:mt-0 md:col-span-7">
                <p class="text-base leading-6 text-neutral-500">
                  Because they're so good at it. Lorem ipsum dolor sit amet consectetur adipisicing
                  elit. Quas cupiditate laboriosam fugiat.
                </p>
              </dd>
            </div>
            <div class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8">
              <dt class="text-base leading-6 font-medium text-neutral-900 md:col-span-5">
                Mittagsdetails immer einblenden
              </dt>
              <dd class="mt-2 md:mt-0 md:col-span-7">
                <p class="text-base leading-6 text-neutral-500">
                  He couldn't see himself doing it. Lorem ipsum dolor sit, amet consectetur
                  adipisicing elit. Eveniet perspiciatis officiis corrupti tenetur. Temporibus ut
                  voluptatibus, perferendis sed unde rerum deserunt eius.
                </p>
              </dd>
            </div>
            <div class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8">
              <dt class="text-base leading-6 font-medium text-neutral-900 md:col-span-5">
                Animationen ausschalten
              </dt>
              <dd class="mt-2 md:mt-0 md:col-span-7">
                <p class="text-base leading-6 text-neutral-500">
                  He couldn't see himself doing it. Lorem ipsum dolor sit, amet consectetur
                  adipisicing elit. Eveniet perspiciatis officiis corrupti tenetur. Temporibus ut
                  voluptatibus, perferendis sed unde rerum deserunt eius.
                </p>
              </dd>
            </div>
          </dl>
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { useTheme } from '@/store/theme'
import { storeToRefs } from 'pinia'

const store = useTheme()
const { themes, currentTheme } = storeToRefs(store)

function onSelectTheme(event: Event) {
  const selectElem = event.target as HTMLSelectElement
  const newTheme = themes.value.find((theme) => theme.cssClass === selectElem?.value)
  if (newTheme) store.setCurrentTheme(newTheme)
}
</script>
