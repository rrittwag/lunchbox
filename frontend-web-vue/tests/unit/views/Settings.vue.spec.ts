import Settings from '@/views/Settings.vue'
import { useTheme, THEME_RED } from '@/store/theme'
import { shallowMount } from '@vue/test-utils'
import { describe, test, expect } from 'vitest'
import { createTestingPinia } from '@tests/unit/test-utils'

describe('Settings', () => {
  test('renders snapshot', () => {
    const pinia = createTestingPinia()
    useTheme().$patch({ currentTheme: THEME_RED, colorScheme: 'system' })

    const wrapper = shallowMount(Settings, {
      global: { plugins: [pinia] },
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
