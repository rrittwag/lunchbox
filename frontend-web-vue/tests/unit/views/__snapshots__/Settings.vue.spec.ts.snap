// Vitest Snapshot v1

exports[`Settings > renders snapshot 1`] = `
<div
  class="sm:p-4"
>
  <h1
    class="flex items-center justify-center px-4 text-2xl text-neutral-800 sm:h-16 sm:justify-start"
  >
     Einstellungen 
  </h1>
  <div>
    <div
      class="mx-auto max-w-screen-xl py-12 px-4 sm:px-6 lg:py-16 lg:px-8"
    >
      <div
        class="mt-6 border-t-2 border-neutral-500 pt-6"
      >
        <dl>
          <div
            class="md:grid md:grid-cols-12 md:gap-8"
          >
            <dt
              class="text-base font-medium leading-6 text-neutral-900 md:col-span-5"
            >
              Ort
            </dt>
            <dd
              class="mt-2 md:col-span-7 md:mt-0"
            >
              <button
                class="h-20 rounded border border-primary-500 px-4 text-base leading-6 text-neutral-800"
              >
                 Neubrandenburg 
              </button>
              <button
                class="ml-2 h-20 rounded border border-neutral-400 px-4 text-base leading-6 text-neutral-500 hover:border-neutral-800"
              >
                 Neubrandenburg 
              </button>
              <button
                class="ml-2 h-20 rounded border border-neutral-400 px-4 text-base leading-6 text-neutral-500 hover:border-neutral-800"
              >
                 Berlin Springpfuhl 
              </button>
            </dd>
          </div>
          <div
            class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8"
          >
            <dt
              class="text-base font-medium leading-6 text-neutral-900 md:col-span-5"
              for="settings-theme"
            >
               Farbvariante 
            </dt>
            <dd
              class="mt-2 md:col-span-7 md:mt-0"
            >
              <div
                class="relative inline-block w-32"
              >
                <select
                  class="block w-full appearance-none rounded border border-neutral-400 bg-neutral-200 px-4 py-2 pr-8 leading-tight text-neutral-900 hover:border-neutral-800 focus:outline-none focus:ring-4"
                  id="settings-theme"
                  name="theme"
                >
                  
                  <option
                    value="theme-red"
                  >
                    Red
                  </option>
                  <option
                    value="theme-green"
                  >
                    Green
                  </option>
                  <option
                    value="theme-blue"
                  >
                    Blue
                  </option>
                  <option
                    value="theme-blue-dark"
                  >
                    Blue (Dark)
                  </option>
                  
                </select>
                <div
                  class="pointer-events-none absolute inset-y-0 right-0 flex items-center px-2 text-neutral-900"
                >
                  <svg
                    class="h-4 w-4 fill-current"
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"
                    />
                  </svg>
                </div>
              </div>
            </dd>
          </div>
          <div
            class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8"
          >
            <dt
              class="text-base font-medium leading-6 text-neutral-900 md:col-span-5"
            >
               Dark Mode 
            </dt>
            <dd
              class="mt-2 md:col-span-7 md:mt-0"
            >
              <p
                class="text-base leading-6 text-neutral-500"
              >
                ja/nein/System
              </p>
            </dd>
          </div>
          <div
            class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8"
          >
            <dt
              class="text-base font-medium leading-6 text-neutral-900 md:col-span-5"
            >
               Mittagsanbieter anzeigen 
            </dt>
            <dd
              class="mt-2 md:col-span-7 md:mt-0"
            >
              <p
                class="text-base leading-6 text-neutral-500"
              >
                 Because they're so good at it. Lorem ipsum dolor sit amet consectetur adipisicing elit. Quas cupiditate laboriosam fugiat. 
              </p>
            </dd>
          </div>
          <div
            class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8"
          >
            <dt
              class="text-base font-medium leading-6 text-neutral-900 md:col-span-5"
            >
               Mittagsdetails immer einblenden 
            </dt>
            <dd
              class="mt-2 md:col-span-7 md:mt-0"
            >
              <p
                class="text-base leading-6 text-neutral-500"
              >
                 He couldn't see himself doing it. Lorem ipsum dolor sit, amet consectetur adipisicing elit. Eveniet perspiciatis officiis corrupti tenetur. Temporibus ut voluptatibus, perferendis sed unde rerum deserunt eius. 
              </p>
            </dd>
          </div>
          <div
            class="mt-8 border-t border-neutral-500 pt-6 md:grid md:grid-cols-12 md:gap-8"
          >
            <dt
              class="text-base font-medium leading-6 text-neutral-900 md:col-span-5"
            >
               Animationen ausschalten 
            </dt>
            <dd
              class="mt-2 md:col-span-7 md:mt-0"
            >
              <p
                class="text-base leading-6 text-neutral-500"
              >
                 He couldn't see himself doing it. Lorem ipsum dolor sit, amet consectetur adipisicing elit. Eveniet perspiciatis officiis corrupti tenetur. Temporibus ut voluptatibus, perferendis sed unde rerum deserunt eius. 
              </p>
            </dd>
          </div>
        </dl>
      </div>
    </div>
  </div>
</div>
`;
